Add documentations for PEFT and save_pretrained option (#11232)
mlflow/mlflow: docs/source/llms/transformers/guide/index.rst

\***** If using `pyfunc` in MLflow Model Serving for realtime inference, the raw audio in bytes format must be base64 encoded prior to submitting to the endpoint. String inputs will be interpreted as uri locations.

> Example of loading a transformers model as a python function

Author: note: not a new section, just moving under the (probably) right section.

    result = pyfunc_loaded.predict(data, params=params)

Pipelines vs. Component Logging
-------------------------------

The transformers flavor has two different primary mechanisms for saving and loading models: pipelines and components.

**Pipelines**

Pipelines in the context of the Transformers library are high-level objects that combine pre-trained models and tokenizers 
(as well as other components, depending on the task type) to perform a specific task. They abstract away much of the preprocessing 
and postprocessing work involved in using the models. 

For example, a text classification pipeline would handle the tokenization of text, passing the tokens through a model, and then 

> interpreting the logits to produce a human-readable classification.

Reviewer: ```suggestion
interpret the logits to produce a human-readable classification.
```

.. note:: Audio models being used for serving that intend to utilize pre-formatted audio in ``np.ndarray`` format
    must have the model saved with a signature configuration that reflects this schema. Failure to do so will result in type casting errors due to the default signature for
    audio transformers pipelines being set as expecting ``binary`` (``bytes``) data. The serving endpoint cannot accept a union of types, so a particular model instance must choose one
    or the other as an allowed input type.

.. _transformers-save-pretrained-guide:

Storage-Efficient Model Logging with ``save_pretrained`` Option
---------------------------------------------------------------

.. warning::

    The ``save_pretrained`` argument is only available in MLflow 2.11.0 and above, and still in experimental stage. The API and behavior may change in future releases. Moreover, this feature is intended for advanced users who are familiar with Transformers and MLflow, understanding :ref:`the potential risks <caveats-of-save-pretrained>` of using this feature.

Avoiding Redundant Model Copy by Setting ``save_pretrained=False``
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> When MLflow logs an ML model, it always saves a copy of the model weight to the artifact store.

Reviewer: ```suggestion
Typically, when MLflow logs an ML model, it always saves a copy of the model weight to the artifact store.
```

However, this is not optimal when you use a pretrained model from HuggingFace Hub as it is i.e. without any modification to the model weight, but just changing other assets such as prompt and parameters. For such a case, those copy of model weight is redundant and takes up unnecessary storage space.

> To address this issue, MLflow 2.11.0 introduced a new argument ``save_pretrained`` in the :py:func:`mlflow.transformers.save_model()` or :py:func:`mlflow.transformers.log_model()` APIs. When it is set to ``False``, MLflow will not save the copy of the pretrained model weight, but only the reference to the HuggingFace Hub, namely, a repository name and a commit hash. When loading back such *refernce-only* model, MLflow will check the repository name and commit hash from the saved metadata, and download the model weight from the HuggingFace Hub, or use the cached model from HuggingFace local cache directory. The ``save_pretrained`` argument is set to ``True`` by default and doesn't change the model saving behavior

Author: Makes sense, will do!

Example Usage
^^^^^^^^^^^^^

Here is the example of using ``save_pretrained`` argument for logging a model

.. code-block:: python

    import transformers

    pipeline = transformers.pipeline(
        task="text-generation", model="databricks/dolly-v2-7b", torch_dtype="torch.float16"
    )

    with mlflow.start_run():
        mlflow.transformers.log_model(
            transformers_model=pipeline,
            artifact_path="dolly",
            save_pretrained=False,
        )

MLflow will not save the copy of Dolly-v2-7B model weight, instead logs the following metadata as a reference to the HuggingFace Hub model. This will save storage space about 15GB and reduce the logging latency significantly as well.
```
source_model_name: "databricks/dolly-v2-7b"
source_model_revision: "d632f0c8b75b1ae5b26b250d25bfba4e99cb7c6f"
```

.. _caveats-of-save-pretrained:

Caveats of Reference-Only Models
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

While the ``save_pretrained`` argument is useful for saving storage space and reducing logging latency, it has the following caveats to be aware of:

> * **Change in Model Unavailability**: If you are using a model from other users' repository, themodel may be deleted or become private in the HuggingFace Hub. In such case, MLflow cannot load the model back. For production use cases, it is recommended to save the copy model weight to the artifact store once for your production model.

Reviewer: ```suggestion
* **Change in Model Unavailability**: If you are using a model from other users' repository, the model may be deleted or become private in the HuggingFace Hub. In such cases, MLflow cannot load the model back. For production use cases, it is recommended to save the copy model weight to the artifact store prior to moving from development or staging to production for your model.
```

> * **HuggingFace Hub Access**: The model downloading from the HuggingFace Hub might be slow or unstable due to the network condition or the HuggingFace Hub service condition. MLflow doesn't provide any retry mechanism or robust error handling for the model downloading.

Reviewer: ```suggestion
* **HuggingFace Hub Access**: Downloading a model from the HuggingFace Hub might be slow or unstable due to the network condition or the HuggingFace Hub service status. MLflow doesn't provide any retry mechanism or robust error handling for the model downloading. As such, you should not rely on this functionality for your final production-candidate run.
```

* **Limited Databricks Integration**: If you are using Databricks, be aware that the model saved with `save_pretrained=False` cannot be registered to the legacy `Workspace Model Registry <https://docs.databricks.com/en/machine-learning/manage-model-lifecycle/workspace-model-registry.html>`_. If you want to register the reference-only Transformer model, please use `Unity Catalog <https://docs.databricks.com/en/machine-learning/manage-model-lifecycle/index.html>`_ instead, or download the model weight in advance using :py:func:`mlflow.transformers.persist_pretrained_model()` API as described in the next section.

.. _persist-pretrained-guide:

Persist the Model Weight to the Existing Reference-Only Model
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> If you want to update the reference-only model to the one contains the model weight, you can use :py:func:`mlflow.transformers.persist_pretrained_model()` API. This API will download the model weight from the HuggingFace Hub, save it to the artifact location, and update the metadata of the given reference-only model. After this operation, the model will be equivalent to the one saved with `save_pretrained=True` and be ready for the production use.

Reviewer: ```suggestion
If you want to update the reference-only model to an instance that contains the model weight, you can use the :py:func:`mlflow.transformers.persist_pretrained_model()` API. This API will download the model weight from the HuggingFace Hub, save it to the artifact location, and update the metadata of the given reference-only model. After this operation, the model will be equivalent to the one saved with `save_pretrained=True` and be ready for the production use.
```
Can we add a callout to be very clear to users that they don't have to re-log a new run in order to update the artifact with the model weights with some nice bold text to REALLY warn them about this?

Author: Great idea! Adding a tip box and stress the fact that the update doesn't require re-logging.

    pipeline = transformers.pipeline(
        task="text-generation", model="databricks/dolly-v2-7b", torch_dtype="torch.float16"
    )

    # Save the reference-only Transformer model
    with mlflow.start_run():
        model_info = mlflow.transformers.log_model(
            transformers_model=pipeline,
            artifact_path="dolly",
            save_pretrained=False,
        )

    # Model weight is not saved to the artifact store
    assert not os.path.exists(model_info.artifact_path + "/model")

    # This will download the model weight from the HuggingFace Hub and save it
    # to the artifact location
    mlflow.transformers.persist_pretrained_model(model_info.model_uri)

    assert os.path.exists(model_info.artifact_path + "/model")

PEFT Models in MLflow Transformers flavor
-----------------------------------------

.. warning::

    The PEFT model is supported in MLflow 2.11.0 and above and is still in the experimental stage. The API and behavior may change in future releases. Moreover, the `PEFT <https://huggingface.co/docs/peft/en/index>`_ library is under active development, so not all features
    and adapter types might be supported in MLflow.

> `PEFT <https://huggingface.co/docs/peft/en/index>`_ is a library developed by HuggingFace🤗, that provides various optimization methods for pretrained models available on the HuggingFace Hub. With PEFT, you can easily apply various optimization techniques like LoRA, QLoRA, to reduce the cost of fine-tuning Transformers models.

Reviewer: ```suggestion
`PEFT <https://huggingface.co/docs/peft/en/index>`_ is a library developed by HuggingFace🤗, that provides various optimization methods for pretrained models available on the HuggingFace Hub. With PEFT, you can easily apply various optimization techniques like LoRA and QLoRA to reduce the cost of fine-tuning Transformers models.
```
Can we add 1-2 sentences that give an overview of what this methodology means in the scope of "accelerated fine tuning" and how it can save applied GenAI developers a TON of time and headache compared with traditional fine tuning techniques? (speed, cost, accuracy, relevancy, hallucination reduction, etc...)

Author: Sounds great! Will add quick description of LoRA and how PEFT make it easy to use.

Format of Saved PEFT Model
^^^^^^^^^^^^^^^^^^^^^^^^^^
When saving PEFT models, MLflow only saves the PEFT adapter and the configuration, but not the base model weight. This is the same behavior as the Transformer's `save_pretrained() <https://huggingface.co/docs/transformers/v4.38.1/en/main_classes/model#transformers.PreTrainedModel.save_pretrained>`_ method and is super efficient in terms of storage space and logging latency. One difference is that MLflow alsot saves the HuggingFace Hub repository name and version for the base model in the model metadata, so that it can load the same base model when loading the PEFT model. Concretely, the following artifacts are saved in MLflow for PEFT models:

* The PEFT adapter weight under the ``/peft`` directory.
* The PEFT configuration as a JSON file under the ``/peft`` directory.
* The HuggingFace Hub repository name and commit hash for the base model in the ``MLModel`` metadata file.

Limitations of PEFT Models in MLflow
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> Since the model saving/loading behavior for PEFT models is similar to that with ``save_pretrained=False``, :ref:`the same caveats <caveats-of-save-pretrained>` apply to PEFT models. For example, the base model weight may be deleted or become private in the HuggingFace Hub, and PEFT models cannot be registered to the legacy Databricks Workspace Model Registry.

Reviewer: ```suggestion
Since the model saving/loading behavior for PEFT models is similar to that of ``save_pretrained=False``, :ref:`the same caveats <caveats-of-save-pretrained>` apply to PEFT models. For example, the base model weight may be deleted or become private in the HuggingFace Hub, and PEFT models cannot be registered to the legacy Databricks Workspace Model Registry.
```
